### PR TITLE
Test icpx instead of icpc

### LINKED ISSUE #15301: github-actions: Switch to Intel ICX compiler.
We need to switch to the new Intel ICX compilers in our github-actions worker later this year.

```
icpc: remark #10441: The Intel(R) C++ Compiler Classic (ICC) is deprecated and will be removed from product release in the second half of 2023. The Intel(R) oneAPI DPC++/C++ Compiler (ICX) is the recommended compiler moving forward. Please transition to use this compiler. Use '-diag-disable=10441' to disable this message.
```

Useful resources for porting:

- [Porting Guide for ICC Users to DPCPP or ICX](https://www.intel.com/content/www/us/en/developer/articles/guide/porting-guide-for-icc-users-to-dpcpp-or-icx.html)
- [Compile and Run MPI Programs Using C++ with SYCL](https://www.intel.com/content/www/us/en/developer/articles/technical/compile-and-run-mpi-programs-using-dpcpp-language.html)
- [Build googletest with IntelLLVM compilers. ](https://github.com/google/googletest/commit/e032d57642d2a042b9445b2199277bb24fee2eb5)

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #15301. We could of course add this as an extra build.

### CODE REVIEW

File: source/base/qprojector.cc [hunk 1]
```diff
@@ -769,11 +769,11 @@ QProjector<2>::project_to_all_faces(const ReferenceCell &     reference_cell,
     {
       const auto support_points_line =
         [](const auto &face, const auto &orientation) -> std::vector<Point<2>> {
-        std::array<Point<2>, 2> vertices;
-        std::copy_n(face.first.begin(), face.first.size(), vertices.begin());
-        const auto temp =
-          ReferenceCells::Line.permute_according_orientation(vertices,
-                                                             orientation);
+        ArrayView<const Point<2>> vertices(face.first.begin(),
```
Comment (bangerth) [inline source/base/qprojector.cc:772]: Nice catch!

File: include/deal.II/grid/grid_reordering.h [hunk 2]
```diff
@@ -35,7 +35,7 @@ DEAL_II_NAMESPACE_OPEN
  * this class. Usage of the old-style numbering is deprecated.
  */
 template <int dim, int spacedim = dim>
-class DEAL_II_DEPRECATED GridReordering
+class GridReordering
```
Comment (bangerth) [inline include/deal.II/grid/grid_reordering.h:38]: ?? Do we use this class anywhere so that we want to un-deprecate it?

Comment (masterleinad) [reply include/deal.II/grid/grid_reordering.h:38]: Ah, forgot to hit send:
All its members are deprecated so this only avoids deprecation warnings when defining those in the `*.cc` file.

**Review by marcfehling — APPROVED** [review]

**Comment by marcfehling** [comment]
The windows tester doesn't like this:
```
D:\a\dealii\dealii\source\base\qprojector.cc(772,43): error C2664: 'dealii::ArrayView<const dealii::Point<2,Number>,dealii::MemorySpace::Host>::ArrayView(const dealii::Point<2,Number> *,const size_t)': cannot convert argument 1 from 'std::_Array_const_iterator<_Ty,2>' to 'const dealii::Point<2,Number> *' [D:\a\dealii\dealii\build\source\base\object_base_debug.vcxproj]
          with
          [
              Number=double
          ]
          and
          [
              _Ty=dealii::Point<2,double>
          ]
          and
          [
              Number=double
          ]
D:\a\dealii\dealii\source\base\qprojector.cc(772,60): message : No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called [D:\a\dealii\dealii\build\source\base\object_base_debug.vcxproj]
D:\a\dealii\dealii\include\deal.II/base/array_view.h(128,3): message : see declaration of 'dealii::ArrayView<const dealii::Point<2,Number>,dealii::MemorySpace::Host>::ArrayView' [D:\a\dealii\dealii\build\source\base\object_base_debug.vcxproj]
          with
          [
              Number=double
          ]
D:\a\dealii\dealii\source\base\qprojector.cc(805): message : see reference to function template instantiation 'std::vector<dealii::Point<2,Number>,std::allocator<dealii::Point<2,Number>>> dealii::QProjector<2>::project_to_all_faces::<lambda_0c83d88963be66288dded01df112f8f9>::operator ()<_Ty,unsigned int>(const _Ty &,const unsigned int &) const' being compiled [D:\a\dealii\dealii\build\source\base\object_base_debug.vcxproj]
          with
          [
              Number=double,
              _Ty=std::pair<std::array<dealii::Point<2,double>,2>,double>
          ]
  fe_dgp_monomial.cc
D:\a\dealii\dealii\source\base\qprojector.cc(804): message : see reference to function template instantiation 'std::vector<dealii::Point<2,Number>,std::allocator<dealii::Point<2,Number>>> dealii::QProjector<2>::project_to_all_faces::<lambda_0c83d88963be66288dded01df112f8f9>::operator ()<_Ty,unsigned int>(const _Ty &,const unsigned int &) const' being compiled [D:\a\dealii\dealii\build\source\base\object_base_debug.vcxproj]
          with
          [
              Number=double,
              _Ty=std::pair<std::array<dealii::Point<2,double>,2>,double>
          ]
```

**Review by marcfehling — APPROVED** [review]
MSVC looks happy now :)

**Pull request merged**